rfcs/proposed/numa_support/create-numa-arenas.md

```diff
@@ -0,0 +1,110 @@
+
+```c++
+  std::vector<tbb::task_arena> arenas = tbb::create_numa_task_arenas();
+  std::vector<tbb::task_group> task_groups(arenas.size());
+```
+
+### Public API
+
+The function has the following signature:
+
+```c++
+// Defined in tbb/task_arena.h
+
+namespace tbb {
+    std::vector<tbb::task_arena> create_numa_task_arenas(
+        task_arena::constraints other_constraints = {},
+        unsigned reserved_slots = 0
+    };
+}
+```
+
+It optionally takes a `constraints` argument to change default arena settings such as maximal concurrency
+(the upper limit on the number of threads), core type etc.; the `numa_id` value in `other_constraints`
+is ignored. The second optional argument allows to override the number of reserved slots, which by default
+is 0 (unlike the `task_arena` construction default of 1) for the reasons described in the introduction.
+
+The function returns a `std::vector` of created arenas. The arenas should not be initialized,
+in order to allow changing certain arena settings before the use.
+
+### Possible implementation
+
+```c++
+std::vector<tbb::task_arena> create_numa_task_arenas(
+    tbb::task_arena::constraints other_constraints,
+    unsigned reserved_slots)
+{
+    std::vector<tbb::numa_node_id> numa_nodes = tbb::info::numa_nodes();
+    std::vector<tbb::task_arena> arenas;
+    arenas.reserve(numa_nodes.size());
+    for (tbb::numa_node_id nid : numa_nodes) {
+        other_constraints.numa_id = nid;
+        arenas.emplace_back(other_constraints, reserved_slots);
+    }
+    return arenas;
+}
+```
+
+## Considered alternatives
+
+The earlier proposal [PR #1559](https://github.com/uxlfoundation/oneTBB/pull/1559) suggested to add
+a new class derived from `task_arena` which would also provide a method to wait for work completion.
+It aims to simplify the whole usage example as much as possible. That proposal also includes
+a factory function to create a set of such arenas constrained by NUMA domains.
+
+While suggesting very concise and "bullet-proof" API for the problem at question, that approach has
+its downsides:
+- Adding a special "flavour" of `task_arena` potentially increases the library learning curve and
+  might create confusion about which class to use in which conditions, and how these interoperate.
+- It seems very specialized, capable to only address specific and quite narrow set of use cases.
+
+Our proposal, instead, aims at a single usability aspect with incremental improvements/extensions
+to the existing oneTBB classes and usage patterns, leaving other aspects to complementary proposals.
+Specifically, the [Waiting in a task arena](../task_arena_waiting/readme.md) RFC improves the joint
+use of a task arena and a task group to submit and wait for work. Combined, these extensions will
+provide only a slightly more verbose solution for the NUMA use case, while being more flexible
+and having greater potential for other useful extensions and applications.
```

Comment:
I would suggest showing a combined example here, since one of the tradeoffs between [PR #1559](https://github.com/uxlfoundation/oneTBB/pull/1559) and this proposal is verbosity vs flexibility. You have already reproduced the motivating example. PR 1559 shows its implementation of that complete motivating example. Showing the combined solutions with this proposal and [Waiting in a task arena](../task_arena_waiting/readme.md) will let us directly see how much it is simplified compared to the motivating example.

Reply:
For now, the example can be found in the discussion of https://github.com/uxlfoundation/oneTBB/pull/1559#discussion_r1842415581 as well as in the arena waiting proposal https://github.com/uxlfoundation/oneTBB/blob/rfc/create-numa-task-arenas/rfcs/proposed/task_arena_waiting/readme.md#1-simplify-the-use-of-a-task-group.

Eventually we need to "merge" #1559 and this PR, describing one approach as the proposed solution and the other as a considered but declined alternative. I have done some initial steps for that here, but before further "merging" I would like to see some indication that this approach is generally preferred. I hope there is sufficient information available for reviewers to judge on that.

Reply:
Since @aleksei-fedotov has supported this proposal, I added more details from #1559 as well as the code examples for comparison.